buchschloss/gui2/widgets.py

```diff
@@ -1,340 +1,370 @@
-"""widgets"""
-import enum
-import functools
+"""specific widgets and form helpers"""
 import operator
 import tkinter as tk
 import tkinter.ttk as ttk
-from tkinter import Entry, Label, Button
+import tkinter.messagebox as tk_msg
 from functools import partial
-from collections import abc
-import typing as T
+from typing import Callable, Any, Optional
 
 from ..misc import tkstuff as mtk
-from ..misc.tkstuff import dialogs as mtkd
-from ..misc.tkstuff.blocks import PasswordEntry, CheckbuttonWithVar
 
-from . import validation
-from . import common
+from .. import core
 from .. import utils
 from .. import config
-from .. import core
+from . import formlib
+from . import common
 
 
-class ListEntryMixin(Entry):
-    def __init__(self, master, cnf={}, sep=';', **kw):
-        self.sep = sep
-        super().__init__(master, cnf, **kw)
+WRAPLENGTH = config.gui2.widget_size.main.width // 2
 
-    def get(self):
-        v = super().get()
-        if v:
-            return v.split(self.sep)
-        else:
-            return []
-
-
-ListEntry = type('ListEntry', (ListEntryMixin, Entry), {})
-ListREntry = type('ListREntry', (ListEntryMixin, mtk.RememberingEntry), {})
-ISBNEntry = mtk.ValidatedWidget.new_cls(Entry, validation.ISBN_validator)
-NonEmptyEntry = mtk.ValidatedWidget.new_cls(Entry, validation.nonempty)
-NonEmptyREntry = mtk.ValidatedWidget.new_cls(mtk.RememberingEntry, validation.nonempty)
-ClassEntry = mtk.ValidatedWidget.new_cls(Entry, validation.class_validator)
-IntListEntry = mtk.ValidatedWidget.new_cls(ListEntry, validation.int_list)
-IntEntry = mtk.ValidatedWidget.new_cls(Entry, validation.type_int)
-NullIntEntry = mtk.ValidatedWidget.new_cls(Entry, validation.int_or_none)
-NullEntry = mtk.ValidatedWidget.new_cls(Entry, validation.none_on_empty)
-NullREntry = mtk.ValidatedWidget.new_cls(mtk.RememberingEntry, validation.none_on_empty)
-ScriptNameEntry = mtk.ValidatedWidget.new_cls(Entry, validation.script_name)
-
-
-class SeriesEntry(mtk.ContainingWidget):
-    """Entry combining a name for the series and an integer input for the number"""
-
-    def __init__(self, master, **kw):
-        self.number_dummy = core.Dummy(set=self.set_number,
-                                       get=lambda: 1,
-                                       validate=lambda: (1, 0))
-        widgets = [
-            (NullREntry, {'rem_key': 'book-series'}),
-            (NullIntEntry, {'width': 2}),
-        ]
-        super().__init__(master, *widgets, **kw)
 
-    def validate(self):
-        """check whether the series number is valid
+class OptionsFromSearch(formlib.DropdownChoices):
```

Comment:
This should probably also allow the ID in ``.get()``, even if there are multiple alternatives (think e.g. 1 ⇒ 1, 1x, x1, ...)